File: 2218.maximum-value-of-k-coins-from-piles.cs
/*
 * @lc app=leetcode id=2218 lang=csharp
 *
 * [2218] Maximum Value of K Coins From Piles
 */

// @lc code=start
using System.Linq;
using System.Collections.Generic;
using System;
public class Solution 
{
    public int MaxValueOfCoins(IList<IList<int>> piles, int k) 
    {
        int n = piles.Count;
        int[][] dp = new int[n + 1][];
        dp = dp.Select(x => Enumerable.Repeat(0, k + 1).ToArray()).ToArray();
        int prefixSum;
        for (int i = 0; i < n; i++)
        {
            prefixSum = 0;
            for (int j = 0; j < piles[i].Count + 1; j++)
            {
                for (int l = 0; l <= k - j; l++)
                {
                    dp[i + 1][l + j] = Math.Max(dp[i + 1][l + j], dp[i][l] + prefixSum);
                }
                if (j < piles[i].Count)
                {
                    prefixSum += piles[i][j];
                }
            }
        }
        return dp[n][k];
    }
}
// @lc code=end

File: 2218.maximum-value-of-k-coins-from-piles.py
from typing import List


class Solution:
    '''
    Think about it in DP. Suppose we only have piles[0]. Then to get k coins (assuming k <= len(piles[0])), we just add piles[0][:k]
    Now suppose we introduce piles[1]. Then we have to consider among k, we take i from piles[1] and k - i piles from piles[:1]
    dp[i][j] =  maximum total value of coins you can have in your wallet if you choose exactly j coins optimally, given you are given piles[:i + 1].
    '''
    def maxValueOfCoins(self, piles: List[List[int]], k: int) -> int:
        n = len(piles)
        dp = {0: 0}
        for i in range(n):
            dpNew = {}
            prefixSum = 0       
            for j in range(len(piles[i]) + 1):
                for key, value in dp.items():
                    if key + j <= k:
                        if key + j not in dpNew:
                            dpNew[key + j] = value + prefixSum
                        else:
                            dpNew[key + j] = max(dpNew[key + j], value + prefixSum)
                if j < len(piles[i]):
                    prefixSum += piles[i][j]
            dp = dpNew
        return dp[k]
    